### PR TITLE
[#8][#9] better structures

### PR DESCRIPTION
**Pull request by StasyaGera**
I left several comments on this code, please answer.

It also seems that I forgot to pull from master beforehand...

### CODE REVIEW

File: src/Todo/Model.hs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+{-# LANGUAGE DeriveGeneric #-}
```
Comment (ChShersh) [inline src/Todo/Model.hs]: This extensions should be added to default extensions since it's completely safe. And it's convenient to not write it manually every time.

File: src/Todo/Model.hs [hunk 2]
```diff
@@ -7,21 +9,40 @@ module Todo.Model
...
+type Todos = HashSet Todo
 
-newtype Title = Title (NonEmpty Char) deriving (Eq, Show)
--- the lower the priority value, the more urgent the task
-newtype Priority = Priority (R.Ratio N.Natural) deriving (Eq, Show, Ord)
-newtype Tag = Tag (NonEmpty Char) deriving (Eq, Show)
-newtype Deadline = Deadline Int deriving (Eq, Show, Ord)
-type Tags = NonEmpty Tag
-type Todos = [Todo]
+instance Hashable State
```
Comment (ChShersh) [inline src/Todo/Model.hs]: Doesn't putting `Hashable` under `deriving` without having `Generic` instance for `newtype` work? This should work for generalized newtype deriving...

File: src/Todo/Model.hs [hunk 3]
```diff
@@ -32,3 +53,33 @@ data Todo = Todo
...
+
+getTopPrio :: Int -> Todos -> Todos
+getTopPrio n = HSet.fromList . take n . sortBy (flip $ comparing todoPriority) . HSet.toList
+
+tagStatistics :: Todos -> [(Tag, Int)]
+tagStatistics = count . concatMap (NESet.toList . todoTags)
+  where
+    -- from Data.List.Unique
+    count :: Ord a => [a] -> [(a, Int)]
+    count = map (\x -> (head x, length x)) . group . sort
```
Comment (ChShersh) [inline src/Todo/Model.hs]: Okay, finally it's time to implement `count` function efficiently using `HashMap` and in more type-safe way. `length` for list is a very slow function. It's better to have mapping from key to counter. Also, `group` function actually has type `[a] -> [NonEmpty a]` so you should use this function from `Data.List.NonEmpty` module:

* https://hackage.haskell.org/package/base-4.11.0.0/docs/Data-List-NonEmpty.html#v:group

File: src/NESet.hs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import Data.HashSet (HashSet)
+import qualified Data.HashSet as HSet
+import Data.Hashable (Hashable (..))
+
+data NonEmptySet a = a :< HashSet a  deriving (Eq, Show)
+
+singleton :: Hashable a => a -> NonEmptySet a
+singleton x = x :< HSet.empty
+
+filter :: (Hashable a, Ord a) => (a -> Bool) -> NonEmptySet a -> NonEmptySet a
```
[the remark this answers is not in the document]
Comment (StasyaGera) [reply src/NESet.hs]: I agree that what you suggest is better, but I just did it similar to NonEmptyList.
I thought maybe they had reasons to do so.

[the remark this answers is not in the document]
Comment (ChShersh) [reply src/NESet.hs]: I see that type of `filter` for `NonEmpty` list is safe:

* https://hackage.haskell.org/package/base-4.11.0.0/docs/Data-List-NonEmpty.html#v:filter

It returns just `[a]`.

**Comment by StasyaGera** [comment]
I guess it has conflicts because you performed the last merge and I forgot to pull changes from master.

File: src/Todo/DB.hs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+data TodoDB = TodoDB
+    { dbTodos :: HashSet Todo
+    , dbTags  :: HashSet Tag
+    } deriving (Eq, Show)
+
+data DBError = SingleTaggedTodo | NotRegisteredTags deriving (Eq, Show)
+
+addTodo :: Todo -> TodoDB -> Either DBError TodoDB
+addTodo todo@Todo {..} TodoDB {..} 
+    | f == mempty = Right TodoDB { dbTodos = HSet.insert todo dbTodos, .. }
```
Comment (ChShersh) [inline src/Todo/DB.hs]: I think `HSet.null f` is slightly better than comparison with `mempty`. Also, variable name `f` is not good since your code is not general enough to have general names. Ans usually `f` is for **f**unctions. I would name it as `missingTags` if you don't mind to have long names. Or just `miss` if you like short names.

File: src/Todo/DB.hs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    | otherwise   = Left NotRegisteredTags
+  where
+    f = NESet.filter (\t -> not $ HSet.member t dbTags) todoTags
+
+addTag :: Tag -> TodoDB -> TodoDB
+addTag tag TodoDB {..} = TodoDB { dbTags = HSet.insert tag dbTags, .. }
+
+deleteTag :: Tag -> TodoDB -> Either DBError TodoDB
+deleteTag tag TodoDB {..} = case mapM removeTag $ toList dbTodos of
```
Comment (ChShersh) [inline src/Todo/DB.hs]: Wow, this is a really cool usage of `mapM`. It's really great that you managed to understand and write such code :+1: I can't understand fast enough what's going on here... But looks cool anyway.

File: src/Todo/DB.hs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    Nothing -> Left SingleTaggedTodo
+    Just x  -> Right TodoDB { dbTodos = HSet.fromList x, ..}
```
Comment (ChShersh) [inline src/Todo/DB.hs]: Probably it's better to name this variable as `todos` to improve code readability.

File: src/Todo/DB.hs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  where
+    removeTag :: Todo -> Maybe Todo
+    removeTag Todo {..} = 
+        (NESet.fromSet $ NESet.delete tag todoTags) >>= \ts -> Just Todo { todoTags = ts, .. }
```
Comment (ChShersh) [inline src/Todo/DB.hs]: Just note for future. Maybe you will rewrite this in future or maybe not. But I think that this code can be written slightly better (in my opinion) using `do`-notation. I didn't tell about it yet. But you can write this code block like this:

```haskell
removeTag :: Todo -> Maybe Todo
removeTag Todo {..} = do 
  ts <- NESet.fromSet $ NESet.delete tag todoTags
  Just Todo { todoTags = ts, .. }
```

Syntax

```haskell
do
  x <- foo
  smth x
```

is just a syntax sugar for

```haskell
foo >>= \x -> smth x
```

File: src/Todo/Model.hs [hunk 9]
```diff
@@ -31,4 +49,32 @@ data Todo = Todo
...
     , todoDeadline    :: Maybe Deadline
     , todoDescription :: Maybe String
-    } deriving (Eq, Show)
+    } deriving (Eq, Show, Generic)
+
+instance Hashable Todo 
+
+defTodo :: Title -> Priority -> Tags -> Todo
+defTodo todoTitle todoPriority todoTags = Todo { todoState = TODO
+                                               , todoSubtasks = HSet.empty
```
Comment (ChShersh) [inline src/Todo/Model.hs]: `HashSet` is an instance of `Monoid` so you can just write `mempty` instead of `HSet.empty`. But if you prefer monomorphic versions it's okay to leave it as it is.

File: todo-cli.cabal [hunk 10]
```diff
@@ -20,12 +20,18 @@ source-repository head
 library
   hs-source-dirs:       src
   build-depends:        base >= 4.9 && < 5
+                      , hashable >= 1.0.1.1 && < 1.3
```
Comment (ChShersh) [inline todo-cli.cabal]: Btw, I just wonder, how do you determine version bounds for `.cabal` file? :slightly_smiling_face:

File: src/Todo/Model.hs [hunk 11]
```diff
@@ -31,4 +49,32 @@ data Todo = Todo
...
+
+tagStatistics :: Todos -> HashMap Tag Int
+tagStatistics = count . concatMap (toList . todoTags)
+  where
+    count :: (Eq a, Hashable a) => [a] -> HashMap a Int
+    count [] = mempty
+    count (x:xs) = HMap.insert x (cnt + 1) mapp
+      where 
+        mapp = count xs
+        cnt = HMap.lookupDefault 0 x mapp
```
Comment (ChShersh) [inline src/Todo/Model.hs]: Yeah, I can understand you here... Because of the fact that `map` is standard function name you need to use names like `mapp` or `mp`. I prefer `dict` as a shorten for _dictionary_ :slightly_smiling_face:

File: src/Todo/Model.hs [hunk 12]
```diff
@@ -31,4 +49,32 @@ data Todo = Todo
...
+
+getTopPrio :: Int -> Todos -> Todos
+getTopPrio n = HSet.fromList . take n . sortBy (flip $ comparing todoPriority) . toList
+
+tagStatistics :: Todos -> HashMap Tag Int
+tagStatistics = count . concatMap (toList . todoTags)
+  where
+    count :: (Eq a, Hashable a) => [a] -> HashMap a Int
+    count [] = mempty
+    count (x:xs) = HMap.insert x (cnt + 1) mapp
```
Comment (ChShersh) [inline src/Todo/Model.hs]: Okay, prepare for `unordered-containers` kung-foo. This library actually has some useful functions.

If you want to insert key with value 1 if it's first key and add 1 to existing value if this is key is already in HashMap you can use [`insertWith`](https://hackage.haskell.org/package/unordered-containers-0.2.9.0/docs/Data-HashMap-Lazy.html#v:insertWith) function.

And you don't need to write pattern-matching on list. You can just use `foldr` function. Unfortuately, your implementation and implementation with `foldr` function is/won't be efficient and tail-recursive. We didn't discuss efficiency in Haskell code yet. But it's better use [`fodl'`](https://hackage.haskell.org/package/base-4.11.0.0/docs/Data-List.html#v:foldl-39-) function.

So, basically, just implement `count` using `foldl'` and `insertWith` functions.

File: src/Todo/Model.hs (hunk 9 above)
Comment (StasyaGera) [reply src/Todo/Model.hs]: It is just a bit more readable to me. If I saw `mempty` in someone else's code I would have to go and check `Monoid` instance for this type (who knows how is it implemented).

File: todo-cli.cabal (hunk 10 above)
Comment (StasyaGera) [reply todo-cli.cabal]: Copied from dependencies list for [this](https://hackage.haskell.org/package/unordered-containers).
